diff --git a/plat/renesas/rz/common/drivers/auth/tbbr/tbbr_cot.c b/plat/renesas/rz/common/drivers/auth/tbbr/tbbr_cot.c
index edf08cb..eaafd98 100644
--- a/plat/renesas/rz/common/drivers/auth/tbbr/tbbr_cot.c
+++ b/plat/renesas/rz/common/drivers/auth/tbbr/tbbr_cot.c
@@ -5,6 +5,7 @@
  */
 
 #include <stddef.h>
+#include <stdint.h>
 
 #include <platform_def.h>
 #include "../../../include/drivers/auth/auth_common.h"
@@ -21,14 +22,14 @@
 
 #ifdef IMAGE_BL2
 
-static unsigned char soc_fw_key_cert_buf[KEY_CERT_LEN];
-static unsigned char soc_fw_content_cert_buf[CONTENT_CERT_LEN];
+static uint32_t soc_fw_key_cert_buf[KEY_CERT_LEN / sizeof(uint32_t)];
+static uint32_t soc_fw_content_cert_buf[CONTENT_CERT_LEN / sizeof(uint32_t)];
 
-static unsigned char tos_fw_key_cert_buf[KEY_CERT_LEN];
-static unsigned char tos_fw_content_cert_buf[CONTENT_CERT_LEN];
+static uint32_t tos_fw_key_cert_buf[KEY_CERT_LEN / sizeof(uint32_t)];
+static uint32_t tos_fw_content_cert_buf[CONTENT_CERT_LEN / sizeof(uint32_t)];
 
-static unsigned char nt_fw_key_cert_buf[KEY_CERT_LEN];
-static unsigned char nt_fw_content_cert_buf[CONTENT_CERT_LEN];
+static uint32_t nt_fw_key_cert_buf[KEY_CERT_LEN / sizeof(uint32_t)];
+static uint32_t nt_fw_content_cert_buf[CONTENT_CERT_LEN / sizeof(uint32_t)];
 
 
 static auth_param_type_desc_t key_cert = AUTH_PARAM_TYPE_DESC(
